src/Confluent.Kafka/Producer.cs

```diff
@@ -664,12 +664,14 @@ internal Producer(ProducerBuilder<TKey, TValue> builder)
             Message<TKey, TValue> message,
             CancellationToken cancellationToken)
         {
+            Headers headers = message.Headers ?? new Headers();
```

Comment:
ensuring headers is never `null` is a better API from the perspective that the typing is more constrained. however:
- this isn't consistent with `Message`, where headers can be `null`.
- it results in an allocation on every produce which in the vast majority of cases is unnecessary. we'd prefer to trade-off usability here imo.

Reply:
@mhowlett Instead of the headers I could change to the entire message. In this case that would be the responsibility of the serializer to allocate the headers if needed.

Reply:
Headers are additive and must not be overwritten: some headers could be added to the Produce() call, others by the serializer, some by producer interceptor plugins and even some by the producer itself.

Reply:
following what @edenhill says, that the `Headers` class should probably really not have a `Remove` method. i think we can leave it as is though - the intent of adding headers here is to make them available to the serdes, not that the serdes should be able to change them.

Reply:
ahh right, i now see you want the headers to be able to be mutated by the serdes (and this feature makes a lot of sense).

i've now thought about this in depth, and think that your original implementation is the best compromise after all. There's an additional allocation, but short lived so it should be negligible. the api downsides seem significantly worse.

you have some white space issues on the `SerializationContext.Headers` comment - (bad indent and space after the 'an'). If you those whitespace issues and revert the 3rd commit, this PR is good to go. lets get it in 1.4!

Reply:
Just push a fix with indentation and space and revert to the right push.
Let me know if it's all good or not. :)

Thank you!